### PR TITLE
OIA-4: Map collectors from integration api

### PR DESCRIPTION
**Pull request by cgorantla**
* JIRA: http://issues.opennms.org/browse/OIA-4

Our [continuous integration system](http://bamboo.internal.opennms.com:8085) will test and verify your changes.

### CODE REVIEW

File: features/collection/shell-commands/src/main/java/org/opennms/netmgt/collection/commands/CollectCommand.java [hunk 1]
```diff
@@ -146,7 +146,7 @@ public Void execute() {
         final CompletableFuture<CollectionSet> future = locationAwareCollectorClient.collect()
                 .withAgent(agent)
                 .withSystemId(systemId)
-                .withCollector(collector)
+                .withCollectorClassName(className)
```
Comment (j-white) [inline features/collection/shell-commands/src/main/java/org/opennms/netmgt/collection/commands/CollectCommand.java]: Is this change necessary? We already use the `ServiceCollectorRegistry` to lookup the collector above.

Comment (j-white) [reply features/collection/shell-commands/src/main/java/org/opennms/netmgt/collection/commands/CollectCommand.java]: I guess it relates to the fact that the CollectorRequestBuilderImpl ends up with the proxy class name instead, if that's it we should add a comment.

File: features/collection/client-rpc/src/main/java/org/opennms/netmgt/collection/client/rpc/CollectorRequestBuilderImpl.java [hunk 2]
```diff
@@ -120,7 +123,7 @@ public CollectorRequestBuilder withAttributes(Map<String, Object> attributes) {
         CollectorRequestDTO request = new CollectorRequestDTO();
         request.setLocation(target.getLocation());
         request.setSystemId(target.getSystemId());
-        request.setClassName(serviceCollector.getClass().getCanonicalName());
+        request.setClassName(className != null ? className : serviceCollector.getClass().getCanonicalName());
```
Comment (j-white) [inline features/collection/client-rpc/src/main/java/org/opennms/netmgt/collection/client/rpc/CollectorRequestBuilderImpl.java]: Can we add a comment explaining why we do this?

File: features/api-layer/src/main/java/org/opennms/features/apilayer/collectors/ServiceCollectorManager.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    public synchronized void onBind(ServiceCollectorFactory extension, Map properties) {
+        LOG.debug("bind called with {}: {}", extension, properties);
+        if (extension != null) {
+            extServiceRegistrationMap.computeIfAbsent(extension, (ext) -> {
+                final ServiceCollector mappedExt = map(ext);
+                final Hashtable<String, Object> props = new Hashtable<>();
+                // Make the service available to any Spring-based listeners
+                props.put("registration.export", Boolean.TRUE.toString());
+                // Registry needs type of collector class.
+                props.put("type", ext.getCollectorClassName());
```
Comment (j-white) [inline features/api-layer/src/main/java/org/opennms/features/apilayer/collectors/ServiceCollectorManager.java]: We should keep the onBind logic in the InterfaceMapper and only extend what we need. Instead of overriding onBind completely I would add something like:
```
    @Override
    public Map<String, Object> getServiceProperties(ServicePollerFactory extension) {
        return ImmutableMap.<String,Object>builder()
                // Registry needs type of collector class.
                .put("type", extension.getCollectorClassName())
                .build();
    }
```

Comment (j-white) [reply features/api-layer/src/main/java/org/opennms/features/apilayer/collectors/ServiceCollectorManager.java]: We can also refactor the `ServicePollerManager` to do the same.

File: features/api-layer/src/main/java/org/opennms/features/apilayer/collectors/ServiceCollectorImpl.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                    org.opennms.netmgt.collection.support.builder.GenericTypeResource genericResource =
+                            new org.opennms.netmgt.collection.support.builder.GenericTypeResource(nodeLevelResource, null, genericTypeResource.getInstance());
+                    addAttributes(collectionSetResource, builder, genericResource);
+                }
+            }
+            builder.withTimestamp(new Date(collectionSet.getTimeStamp()));
+            return builder.build();
+
+        } catch (InterruptedException | ExecutionException e) {
+            LOG.error("collect failed with ", e);
```
Comment (j-white) [inline features/api-layer/src/main/java/org/opennms/features/apilayer/collectors/ServiceCollectorImpl.java]: We should probably rethrow this as a CollectionException instead of just returning `CollectionStatus.UNKNOWN`.

File: features/api-layer/src/main/java/org/opennms/features/apilayer/collectors/ServiceCollectorImpl.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+                    addAttributes(collectionSetResource, builder, nodeLevelResource);
+                } else if (resource.getResourceType().equals(Resource.Type.INTERFACE)) {
+                    IpInterfaceResource ipResource = (IpInterfaceResource) resource;
+                    NodeLevelResource nodeLevelResource = new NodeLevelResource(ipResource.getNodeResource().getNodeId());
+                    InterfaceLevelResource interfaceLevelResource = new InterfaceLevelResource(nodeLevelResource, ipResource.getInstance());
+                    addAttributes(collectionSetResource, builder, interfaceLevelResource);
+                } else if (resource.getResourceType().equals(Resource.Type.GENERIC)) {
+                    GenericTypeResource genericTypeResource = (GenericTypeResource) resource;
+                    NodeLevelResource nodeLevelResource = new NodeLevelResource(genericTypeResource.getNodeResource().getNodeId());
+                    // TODO: Get ResourceType from ResourceTypesDao
```
Comment (j-white) [inline features/api-layer/src/main/java/org/opennms/features/apilayer/collectors/ServiceCollectorImpl.java]: Could also try using the `DeferredGenericTypeResource` to avoid having to lookup the types.

**Comment by cgorantla** [comment]
Thanks @j-white  for the review. Moved mapper to different class and added unit test for mapper.
